lib/spack/spack/util/s3.py

```diff
@@ -31,8 +31,16 @@ def create_s3_session(url):
     from boto3 import Session
     from botocore.exceptions import ClientError
 
-    session = Session()
+    s3_connection = {}
 
+    if connection:
+        if connection['token']:
+            s3_connection["aws_session_token"] = connection["token"]
+        if connection["pair"][0]:
+            s3_connection["aws_access_key_id"] = connection["pair"][0]
+            s3_connection["aws_secret_access_key"] = connection["pair"][1]
+
+    session = Session(**s3_connection)
     s3_client_args = {"use_ssl": spack.config.get('config:verify_ssl')}
 
     endpoint_url = os.environ.get('S3_ENDPOINT_URL')
```

Comment:
I don't mean to derail the PR, as it's looking pretty good otherwise, but I recall that the endpoint url was another relevant piece of connection information, and it seems like this PR would be an obvious place to include it, along with the keys and session token.

For context, this is an environment variable that you can set to access s3-compliant endpoints other than AWS.

Would be nice to see support for this field added, but this is really an FYI, more than anything else.